src/dune_lang/package.ml

```diff
@@ -121,6 +121,36 @@ let decode =
         ; Pp.textf "- %s" (print_value loc2)
         ]
   in
+  let check_duplicate_deps deps field_name =
+    ignore
+      (List.fold_left deps ~init:[] ~f:(fun seen (loc, dep) ->
+         (match
+            List.find_opt
+              ~f:(fun (_, seen_dep) ->
+                Package_name.equal
+                  dep.Package_dependency.name
+                  seen_dep.Package_dependency.name)
+              seen
+          with
+          | Some _ ->
+            let dep_sexp = Package_dependency.encode dep in
+            let dep_string = Dune_sexp.to_string dep_sexp in
+            User_warning.emit
+              ~loc
+              ~hints:
+                [ Pp.text
+                    "Duplicate dependencies on the same package are redundant. If you \
```

Comment:
This hint probably doesn't need to reiterate the error message. The hint could also be a bit more helpful for those users who don't know about `(and)`. We can of course actually construct what they should have written here instead, but that's up to you. Maybe worth just leaving a CR for now.